contracts/src/CrossChainApplications/NativeTokenBridge/NativeTokenDestination.sol

```diff
@@ -0,0 +1,167 @@
+
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+import "@subnet-evm-contracts/interfaces/IWarpMessenger.sol";
+import "@subnet-evm-contracts/interfaces/IAllowList.sol";
+import "@subnet-evm-contracts/interfaces/INativeMinter.sol";
+import "./INativeTokenDestination.sol";
+import "../../Teleporter/ITeleporterMessenger.sol";
+import "../../Teleporter/ITeleporterReceiver.sol";
+import "../../Teleporter/SafeERC20TransferFrom.sol";
+
+// Precompiled Native Minter Contract Address
+address constant MINTER_ADDRESS = 0x0200000000000000000000000000000000000001;
+// Designated Blackhole Address
+address constant BLACKHOLE_ADDRESS = 0x0100000000000000000000000000000000000000;
+
+contract NativeTokenDestination is
+    ITeleporterReceiver,
+    INativeTokenDestination,
+    ReentrancyGuard
+{
+    INativeMinter private immutable _nativeMinter =
+        INativeMinter(MINTER_ADDRESS);
+
+    address public constant WARP_PRECOMPILE_ADDRESS =
+        0x0200000000000000000000000000000000000005;
+
+    uint256 public constant TRANSFER_NATIVE_TOKENS_REQUIRED_GAS = 10_000_000; // TODO this is a placeholder
+    bytes32 public immutable currentChainID;
+    bytes32 public immutable sourceChainID;
+    address public immutable sourceContractAddress;
+
+    // Used for sending an receiving Teleporter messages.
+    ITeleporterMessenger public immutable teleporterMessenger;
+
+    error InvalidTeleporterMessengerAddress();
+    error InvalidRecipientAddress();
+    error InvalidSourceChain();
+    error InvalidPartnerContractAddress();
+    error CannotBridgeTokenWithinSameChain();
+    error Unauthorized();
+    error InsufficientPayment();
+    error InvalidSourceContractAddress();
+    error InsufficientAdjustedAmount(uint256 adjustedAmount, uint256 feeAmount);
+
+    // TODO we probably want to add the original token supply from this chain to the constructor.
+    constructor(
+        address teleporterMessengerAddress,
+        address sourceContractAddress_,
+        bytes32 sourceChainID_
+    ) {
+        currentChainID = WarpMessenger(WARP_PRECOMPILE_ADDRESS)
+            .getBlockchainID();
+
+        if (teleporterMessengerAddress == address(0)) {
+            revert InvalidTeleporterMessengerAddress();
+        }
+        teleporterMessenger = ITeleporterMessenger(teleporterMessengerAddress);
+
+        if (sourceContractAddress_ == address(0)) {
+            revert InvalidSourceContractAddress();
+        }
+        sourceContractAddress = sourceContractAddress_;
+
+        if (sourceChainID_ == currentChainID) {
+            revert CannotBridgeTokenWithinSameChain();
+        }
+        sourceChainID = sourceChainID_;
+    }
+
+    /**
+     * @dev See {ITeleporterReceiver-receiveTeleporterMessage}.
+     *
+     * Receives a Teleporter message and routes to the appropriate internal function call.
+     */
+    function receiveTeleporterMessage(
+        bytes32 nativeChainID,
+        address nativeBridgeAddress,
```

Comment:
Nit: Can we renaming these `senderChainID` and `senderAddress`? Theoretically, anyone can send a message that will this method, but we only want to allow one specific sender (which is the source chain contract)